### ISSUE TITLE
Impossible to use github pull-request commands with github enterprise

### ISSUE THREAD

**Issue by Khady** [magit/magit#3572]
version: Magit 20180815.2257, Git 2.18.0, Emacs 25.1.1, gnu/linux

I tried to configure magit to work with github enterprise today. The
goal is to get functions like `magit-branch-pull-request` to work. The
configuration I wrote is based on those two pages:

- https://magit.vc/manual/ghub/Getting-Started.html#Getting-Started
- https://mrdias.com/2018/04/01/using-magit-with-github-enterprise.html

The instance of github enterprise is on a domain like git.foo.com.

My git configuration is the following:

```
[github]
user = louis
host = "git.foo.com"
[github "git.foo.com"]
user = louis
```

My `~/.authinfo` info file:

```
machine git.foo.com/api/v3 login louis^magit password ...
machine git.foo.com login louis^magit password ...
machine api.github.com login khady^magit password ...
```

I open a file in the git repository and then execute M-x
magit-branch-pull-request or b Y from the magit status buffer. It
prints a long html error in messages. The interesting parts are probably:

```
mapcar: Wrong type argument: listp, "<!DOCTYPE html>
[...]
    You can't perform that action at this time.
[...]
```

`toggle-debug-on-error` tells me it comes from `magit-read-pull-request`.

In the blog post above, the person has `/api/v3` at the end of the
host. So I tried this too.

```
[github]
user = louis
host = "git.foo.com/api/v3"
[github "git.foo.com/api/v3"]
user = louis
```

This time `magit-branch-pull-request` tells me `Currently only Github
is supported` and `b Y` tells me `magit-upstream-repository: Currently
only Github is supported`.

I traced this message to `magit--github-url-p`.

If I change this function to `(defun magit--github-url-p (url) t)`,
then `magit-branch-pull-request` works.

The url parameter of this function is something like
`git@git.foo.com:foo/project.git`. The host is `git.foo.com`. Which
doesn't match `github.com` or `(ghub--host)` (that is
`"git.foo.com/api/v3"`).

Is it considered a bug or I did something wrong with my configuration?

I'm sorry if this bug report is not written in a very formal way and doesn't 
respect the expected format. Thanks for all your work on magit.

**Comment by tarsius** [comment]
I'll be gone for a week and still have to pack, so I'll have to be brief.

`git.foo.com/api/v3` is probably correct from what I have heard from people who have access to an enterprise instance, but someone at your company should know for sure. Whatever the correct value is, you have to use it in both `.authsource` and `git/config`.

> The url parameter of this function is something like
> `git@git.foo.com:foo/project.git`. The host is `git.foo.com`. Which
> doesn't match `github.com` or `(ghub--host)` (that is
> `"git.foo.com/api/v3"`).

Sounds like `git config github.host git.foo.com/api/v3` would fix it.

**Comment by tarsius** [comment]
That should do it.

By the way, all of `magit-collab.el` is going away soonish. There is a new implementation in the [forge](https://github.com/magit/forge) package.

That implementation doesn't take part in the guessing game and shouldn't have a similar bug. However once that is released (and the old implementation removed from Magit), you will have to tell Forge explicitly about the enterprise instance:

```lisp
(push ("git.foo.com" "git.foo.com/api/v3"
       "git.foo.com" forge-github-repository)
      forge-alist)
```

**Comment by Khady** [comment]
The commit above doesn't solve the issue. I still get the `Currently only Github is supported` message.

Some debug information:

```
url -> git@git.foo.com:foo/project.git
(ghub--host) -> git.foo.com/api/v3
(concat host "/" path) -> git.foo.com/foo/project
```

`git.foo.com/api/v3` and `git.foo.com/foo/project` don't match.

So I changed the patch to:

```elisp
;; Needed for Enterprise instances.  #3572
(string-match-p (regexp-quote (ghub--host))
  (concat host "/api/v3"))
```

After that `magit-branch-pull-request` can run a bit further. But still without full success. It gets to this message:

```Git variable `nil' is unset.  Set to: ```

It seems to come from this function: https://github.com/magit/ghub/blob/4d7a0921067cbf3c80d1cb682095bb3a90db55be/ghub.el#L770-L790

When it is called, the argument `host` is `git.foo.com/api/v3` and forge is `nil`. So I added a case in the `cond`, line 782:

```
((eq forge nil)        (format "github.%s.user"    host))
```

This is obviously not a proper fix. But after that `magit-branch-pull-request` is able to run properly. I'm sorry I could not manage to find where `ghub--username` is called from.

My git configuration is still using `git.foo.com/api/v3`:

```
[github]
user = louis
host = "git.foo.com/api/v3"
[github "git.foo.com/api/v3"]
user = louis
```
